File: leetcode/linkedlist_cycle_II.js
// /**
//  * Definition for singly-linked list.

function ListNode(val) {
    this.val = val;
    this.next = null;
}

function print(head) {
  if (!head) return '';
  if (!head.next) return `${head.val}`;
  return `${head.val}->${print(head.next)}`;
}

function CSVtoSLL(s) {
  let arr = s.split(',');
  let val = arr.shift();
  let head = new ListNode(val);
  let prev = head;
  while ( val = arr.shift() ) {
    let cur = new ListNode(val);
    prev.next = cur;
    prev = cur;
  }
  return head;
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
  if (!head || !head.next) return false;
  let r1 = head;
  let r2 = r1.next;
  while (r2 = r2.next) {
    if (!r2.next) break;
    r1 = r1.next;
    r2 = r2.next;
    if (r1 === r2) return true;
  }
  return (r1 === r2);
};


let goodList = CSVtoSLL('3,5,5,68,42,1');
console.log(print(goodList));

File: leetcode/linkedlist_cycle_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None
    def __str__(self):
        outp = []
        nextnode = self
        while nextnode:
            outp.append(nextnode.val)
            nextnode = nextnode.next
        return '->'.join(outp)



class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # self.printSLL(head)
        prev_node_mappings = dict()
        prev = head
        if not head.next:
            return None
        prev_node_mappings[id(prev)] = None
        cur = prev.next
        while cur:
            pkey = id(prev)
            ckey = id(cur)
            if ckey in prev_node_mappings:
                return cur
            else:
                prev_node_mappings[ckey] = pkey
            cur = cur.next
        return None

    # def printSLL(self, node):
    #     outp = []
    #     while node:
    #         outp.append(str(node.val))
    #         node = node.next
    #         if len(outp) > 40:
    #             break
    #     return '->'.join(outp)

